src/rai/rai/tools/ros/debugging.py

```diff
@@ -0,0 +1,199 @@
+    try:
+        timer.start()
+        stdout, stderr = proc.communicate()
+        return stdout, stderr
+    finally:
+        timer.cancel()
+
+
+def run_command(cmd: List[str], timeout: int = 5):
+    # Validate command safety by checking for shell operators
+    # Block potentially dangerous characters
+    if any(char in " ".join(cmd) for char in FORBIDDEN_CHARACTERS):
+        raise ValueError(
+            "Command is not safe to run. The command contains forbidden characters."
+        )
+    stdout, stderr = run_with_timeout(cmd, timeout)
+    output = {}
+    if stdout:
+        output["stdout"] = stdout.decode("utf-8")
+    else:
+        output["stdout"] = "Command returned no stdout output"
+    if stderr:
+        output["stderr"] = stderr.decode("utf-8")
+    else:
+        output["stderr"] = "Command returned no stderr output"
+    return str(output)
+
+
+@tool
+def ros2_action(
+    command: Literal["info", "list", "type"],
+    action_name: str = "",
+    timeout: int = 5,
+):
+    """Run a ROS2 action command
+    Args:
+        command: The action command to run (info/list/type)
+        action_name: Name of the action (required for info and type)
+        timeout: Command timeout in seconds
+    """
+    if command in ["info", "type"]:
+        if not action_name:
+            raise ValueError("Action name is required for info and type commands")
+
+    cmd = ["ros2", "action", command]
+    if action_name:
+        cmd.append(action_name)
+    return run_command(cmd, timeout)
+
+
+@tool
+def ros2_service(
+    command: Literal["call", "find", "info", "list", "type"],
+    service_name: str = "",
+    timeout: int = 5,
+):
+    """Run a ROS2 service command
+    Args:
+        command: The service command to run
+        service_name: Name of the service (required for call, info, and type)
+        timeout: Command timeout in seconds
+    """
+    if command in ["call", "info", "type"]:
+        if not service_name:
+            raise ValueError(
+                "Service name is required for call, info, and type commands"
+            )
+
+    cmd = ["ros2", "service", command]
+    if service_name:
+        cmd.append(service_name)
+    return run_command(cmd, timeout)
+
+
+@tool
+def ros2_node(command: Literal["info", "list"], node_name: str = "", timeout: int = 5):
+    """Run a ROS2 node command
+    Args:
+        command: The node command to run
+        node_name: Name of the node (required for info)
+        timeout: Command timeout in seconds
+    """
+    if command == "info":
+        if not node_name:
+            raise ValueError("Node name is required for info command")
+
+    cmd = ["ros2", "node", command]
+    if node_name:
+        cmd.append(node_name)
+    return run_command(cmd, timeout)
+
+
+@tool
+def ros2_param(
+    command: Literal["delete", "describe", "dump", "get", "list", "set"],
+    node_name: str = "",
+    param_name: str = "",
+    timeout: int = 5,
+):
```

Comment:
I think `ros2 param set` might not supported, because it requires (at least) 3 arguments: 
- node_name
- parameter_name 
- value

```
ros2 param set --help
usage: ros2 param set [-h] [--spin-time SPIN_TIME] [-s] [--no-daemon] [--include-hidden-nodes] [--timeout N] node_name parameter_name value
```

Reply:
Fixed